File: branches/tiles/src/items.c
/*
    items.c - air cans, med packs, batteries, gold coins, etc..

    Copyright 2008 Ido Yehieli

    This file is part of CryptRover.

    CryptRover is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CryptRover is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CryptRover.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <string.h>
#include <stdlib.h>
#include "map.h"
#include "utils.h"
#include "entities.h"
#include "items.h"
#include "io.h"

#ifdef __SDL__
Mix_Chunk *money = NULL;
Mix_Chunk *clunk = NULL;
#endif

void init_items(void) {
#ifdef __SDL__
	if(!money)
		money = Mix_LoadWAV("media/money.wav");
	if(!clunk)
		clunk = Mix_LoadWAV("media/clunk.wav");
#endif
	memset(item_m,(int)NULL,sizeof(item_t *)*Y_*X_);

	for (int i=0; i<ITEMS_; i++) {
		item_t *ci=&item_l[i];
		do {
			ci->y=rand()%Y_;
			ci->x=rand()%X_;
		} while (WALL==tile_m[ci->y][ci->x].type ||
		         NEXT_LEVEL==tile_m[ci->y][ci->x].type ||
		         ent_m[ci->y][ci->x] ||
		         item_m[ci->y][ci->x]);
		if (i<MED_NUM) {
			ci->type=MED_PACK;
			ci->color=C_MED;
		} else if (i-MED_NUM<AIR_NUM) {
			ci->type=AIR_CAN;
			ci->color=C_AIR;
		} else if (i-MED_NUM-AIR_NUM<COIN_NUM) {
			ci->type=BATTERY;
			ci->color=C_BAT;
		} else {
			ci->type=COIN;
			ci->color=C_COIN;
		}
		ci->used=false;

		item_m[ci->y][ci->x]=ci;
	}
}

void use_item(ent_t *pl) {
	item_t *ci = item_m[pl->y][pl->x];
	if (NULL!=ci && !ci->used) {
		if (MED_PACK==ci->type) {
			if (pl->hp<PLAYER_HP) {
				//heal hp
				pl->hp=min(pl->hp+MED_CHARGE,PLAYER_HP);
				ci->used=true;
#ifdef __SDL__
				Mix_PlayChannel(-1, clunk, 0);
#endif
				add_message("You feel healthy.",ci->color);
			} else
				add_message("A med pack.",0);
		} else if (AIR_CAN==ci->type) {
			if (pl->air<PLAYER_AIR) {
				//replenish air
				pl->air=min(pl->air+AIR_CHARGE,PLAYER_AIR);
				ci->used=true;
#ifdef __SDL__
				Mix_PlayChannel(-1, clunk, 0);
#endif
				add_message("You replenish your air supply.",ci->color);
			} else
				add_message("An air cannister.",0);
		} else if (BATTERY==ci->type) {
			if (pl->battery<PLAYER_BATTERY) {
				//charge battery
				pl->battery=min(pl->battery+BATTERY_CHARGE,PLAYER_BATTERY);
				ci->used=true;
#ifdef __SDL__
				Mix_PlayChannel(-1, clunk, 0);
#endif
				add_message("You charge your battery.",ci->color);
			} else
				add_message("A battery.",0);
		} else if (COIN==ci->type) {
			//take coin
			pl->coins+=COIN_CHARGE;
			ci->used=true;
#ifdef __SDL__
				Mix_PlayChannel(-1, money, 0);
#endif
			add_message("You've found a gold coin.",ci->color);
		}
	}
}

File: src-ts/items.ts
// CryptRover Web - Item management

import {
  Y_, X_, WALL, NEXT_LEVEL, ITEMS_,
  MED_PACK, AIR_CAN, BATTERY, COIN,
  MED_NUM, AIR_NUM, BATTERY_NUM,
  MED_CHARGE, AIR_CHARGE, BATTERY_CHARGE, COIN_CHARGE,
  PLAYER_HP, PLAYER_AIR, PLAYER_BATTERY,
  Colors
} from './constants.js';
import { type Item, type Ent, makeGrid } from './types.js';
import { tileM } from './map.js';
import { entM } from './entities.js';
import { randInt, min } from './utils.js';
import { addMessage } from './io.js';
import { playSound } from './audio.js';

// Global item state
export let itemL: Item[] = [];
export let itemM: (Item | null)[][] = [];

// Initialize items for a level
export function initItems(): void {
  // Clear item placement map
  itemM = makeGrid(Y_, X_, () => null);

  // Create item list
  itemL = [];

  for (let i = 0; i < ITEMS_; i++) {
    // Find valid position (not on wall, stairs, entity, or other item)
    let y: number;
    let x: number;
    do {
      y = randInt(0, Y_ - 1);
      x = randInt(0, X_ - 1);
    } while (
      tileM[y][x].type === WALL ||
      tileM[y][x].type === NEXT_LEVEL ||
      entM[y][x] !== null ||
      itemM[y][x] !== null
    );

    // Determine item type based on index
    let glyph: string;
    let type: string;
    let color: string;

    if (i < MED_NUM) {
      glyph = MED_PACK;
      type = 'med';
      color = Colors.med;
    } else if (i < MED_NUM + AIR_NUM) {
      glyph = AIR_CAN;
      type = 'air';
      color = Colors.air;
    } else if (i < MED_NUM + AIR_NUM + BATTERY_NUM) {
      glyph = BATTERY;
      type = 'battery';
      color = Colors.battery;
    } else {
      glyph = COIN;
      type = 'coin';
      color = Colors.coin;
    }

    const item: Item = {
      x,
      y,
      glyph,
      type,
      color,
      used: false,
    };

    itemL.push(item);
    itemM[y][x] = item;
  }
}

// Use item at player's position (auto-consume)
export function useItem(player: Ent): void {
  const item = itemM[player.y][player.x];

  if (item === null || item.used) {
    return;
  }

  switch (item.type) {
    case 'med':
      if (player.hp < PLAYER_HP) {
        player.hp = min(player.hp + MED_CHARGE, PLAYER_HP);
        item.used = true;
        itemM[player.y][player.x] = null;
        addMessage('You feel healthy.', 'success');
        playSound('clunk');
      } else {
        addMessage('A med pack.', '');
      }
      break;

    case 'air':
      if (player.air < PLAYER_AIR) {
        player.air = min(player.air + AIR_CHARGE, PLAYER_AIR);
        item.used = true;
        itemM[player.y][player.x] = null;
        addMessage('You replenish your air supply.', 'info');
        playSound('clunk');
      } else {
        addMessage('An air canister.', '');
      }
      break;

    case 'battery':
      if (player.battery < PLAYER_BATTERY) {
        player.battery = min(player.battery + BATTERY_CHARGE, PLAYER_BATTERY);
        item.used = true;
        itemM[player.y][player.x] = null;
        addMessage('You charge your battery.', 'info');
        playSound('clunk');
      } else {
        addMessage('A battery.', '');
      }
      break;

    case 'coin':
      player.coins += COIN_CHARGE;
      item.used = true;
      itemM[player.y][player.x] = null;
      addMessage("You've found a gold coin.", 'success');
      playSound('money');
      break;
  }
}
